### ISSUE TITLE
Azuracast reverts to default installation message

### ISSUE THREAD

**Issue by pascalla** [AzuraCast/AzuraCast#3978]
**Using Docker installation method**
Yes

**AzuraCast version**
Rolling Release #10f6ba8 (2021-04-03 14:04) • Docker • PHP 7.4

**Host Operating System**
Ubuntu 18.04

**Describe the bug**
The stream seems to revert to the beautifully voice default Azuracast installation message. 

**To Reproduce**
Not sure of root cause.

**Expected behavior**
No Azuracast default installation message

**Relevant Logs**
So the first sign of something odd is these logs out of nowhere:

![image](https://user-images.githubusercontent.com/22846433/113706687-e4291f00-96d6-11eb-8730-a33b3088c340.png)

My current theory that something has eaten up all CPU/RAM and caused everything to stop working. As the following logs are:

![image](https://user-images.githubusercontent.com/22846433/113706987-2f433200-96d7-11eb-835b-907c2cc38efb.png)

These logs continue for a while before fixing themselves

![image](https://user-images.githubusercontent.com/22846433/113707075-4d109700-96d7-11eb-9d46-a429f07cd5dc.png)
 
The MYSQL server dies later on unprovoked

![image](https://user-images.githubusercontent.com/22846433/113707253-8ba65180-96d7-11eb-88be-fda8c5818b89.png)

Looking at slow queries nothing really springs out at me, all seem reasonable for the size of station.

![image](https://user-images.githubusercontent.com/22846433/113707399-c14b3a80-96d7-11eb-99c5-2620a3dd9a90.png)

**Additional context**
This is happening at off-peak times, and seems to be fine during peak times.  

We have just updated from an older stable release, (i'll try find exact commit if useful)

I will probably be updating this ticket with more things i find

**Comment by SlvrEagle23** [comment]
@pascalla Really not sure what caused this currently. The `MySQL server has gone away` error often indicates that the system is out-of-memory (OOM) killing processes; making swap space (if you don't already have any on that server) could help with that.

It just looks like something is completely hammering the CPU/RAM of the server and that's causing it to fallback without much explanation as to why it's doing so.

**Comment by pascalla** [comment]
@SlvrEagle23 

I will allocate swap space, and if it fixes the issue will get more memory on server, current 4gb for only Azuracast.  Will see over few coming days if this fixes it.

 (DigitalOcean keep having scary warnings on allocation swapspace on SSD servers) https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-18-04

**Comment by pascalla** [comment]
After adding 2gb of swap memory, the issue has just occured again

**Comment by SlvrEagle23** [comment]
@pascalla Yeah, it seems _something_ is completely thrashing the system. It doesn't seem to have any particular timing that would coincide with one of our synchronized tasks (at least as far as I can tell, otherwise it would tend to happen at the top of the hour).

If you happen to be SSH'ed into the server when it happens, it would be worthwhile to look at something like `htop` to see what's consuming so many resources. 

I don't see anything in the logs that suggests AzuraCast is doing something particular at that time, but if we can find and pin down what process is causing the problem, we can definitely go from there.

One potential culprit is Liquidsoap 1.4.4, since we did upgrade to that between the version your station was last on and this version, though we haven't had multiple reports of this happening across installations.

As another question, is this happening at all on your test installation, or only on the production one?

**Comment by pascalla** [comment]
Only on production installation,

Currently station is in a weird state where live DJ is still on air, but the rest of Azura seems to be dead. Can not access any of the web interface. Getting a 502 error. (This suggests liquidsoap is out of the question)

These logs are back as expected:

![image](https://user-images.githubusercontent.com/22846433/113861134-c45d2e00-979e-11eb-8736-614572d3f40d.png)

It seems like everything is here:

![image](https://user-images.githubusercontent.com/22846433/113861218-ddfe7580-979e-11eb-9bc5-1d235aed89f9.png)

All the docker containers seem to be running
![image](https://user-images.githubusercontent.com/22846433/113861498-333a8700-979f-11eb-8a0f-ee0e92c49cf4.png)

**Comment by SlvrEagle23** [comment]
@pascalla Nothing in the logs there is really unusual, and it doesn't look like the CPU load is particularly high either.

I'm really not sure what's causing the problem in this case unfortunately. If you are able to find any insight into the issue, please let me know, especially if it's something we can help with on our end.

**Comment by SlvrEagle23** [comment]
@pascalla I've kept looking at the routine synchronization tasks to identify places where the tasks would cause CPU or RAM overruns, and have heavily optimized two separate areas: the registration of listeners and the building of AutoDJ queue items, both of which were routine synchronization tasks that happened every 15 seconds and were thus more prone to causing errors.

I'd highly recommend updating as soon as possible to the latest Rolling Release version, as it should have notable improvements in overall stability and performance, especially for installations with larger media libraries and listener counts.

**Comment by pascalla** [comment]
Appreciate you looking into this, I turned off IP Based statistics and haven't had a crash since, at somepoint today or tomorrow i'll update to latest release and turn back on IP based statistics and hope for the best 👍

**Comment by pascalla** [comment]
Ok i've updated to latest commits after a very odd "crash" with IP statistics turned off. I was unable to login to SSH to be able to see what was using the load. 

![image](https://user-images.githubusercontent.com/22846433/114542401-cb2cea80-9c4f-11eb-93bf-846b6c9fbd56.png)

**Comment by pascalla** [comment]
I moved server (also using a backup) and updated to latest commits, I've yet to experience a crash which is looking good! Really do appreciate the work done. I will be keeping ticket open just for a few more days just confirm!

**Comment by SlvrEagle23** [comment]
@pascalla Great to hear. We've been continuing to optimize where we can, and are indeed still working on some optimizations for things like the media browser, but hopefully the existing optimizations significantly improve your station's performance.

**Comment by pascalla** [comment]
So had a few crashes after a few days of not having any, Found a bit more meaningful logs
![image](https://user-images.githubusercontent.com/22846433/115111691-620fe480-9f79-11eb-83e5-e9a90357baae.png)

Slow Query log doesnt show anything
![image](https://user-images.githubusercontent.com/22846433/115111699-6e943d00-9f79-11eb-9039-1b36e73c84b5.png)

Before the crash has a suspicous delete :
![image](https://user-images.githubusercontent.com/22846433/115111747-af8c5180-9f79-11eb-89a1-a2d7f8591b15.png)

Also maybe this needs seperate issue but I have IP based statistics turned off, however

"[2021-04-17T11:33:50.756944+00:00] AzuraCast.ERROR: NowPlaying adapter error: cURL error 28: Operation timed out after 3000 milliseconds with 45170 out of 97336 bytes received (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://***/admin/listclients?mount=%2Fradio [] {"station":{"id":1,"name":"**"}}" it seems to still be trying to retrieve all clients?

**Comment by SlvrEagle23** [comment]
@pascalla You have IP-based statistics turned off entirely and it's still messing up?

The semaphore thing is another one of those "general signs of something wrong with the overall load of the server" issues...

We're working on adding some profiling tools that will help identify both time and memory hogs across the application currently, and hopefully this will have long-term improvements in the system's performance overall.

I wish I knew what specifically was knocking your station down (short of possibly attacks from external sources), but I don't know what it is or why it's still having issues.

**Comment by SlvrEagle23** [comment]
Update on this issue: I worked with @pascalla directly to inspect their installation and discovered that there was a big issue with database lockups being caused by having multiple worker processes all hitting the same DB table and trying to lock the same rows.

As a fix for this issue, in Rolling Release we limited the number of parallel workers to 1, which appears to have largely resolved the issue in its entirety; we've heard from other stations that the same fix applied on their installations resulted in the same improvement.

The next stable release will incorporate this limitation as well.
